### ISSUE TITLE
Symbols text/icon pointing to the actual point/poi on map in pitched mode (Parity with Google Maps)

### ISSUE THREAD

**Issue by mb12** [mapbox/mapbox-gl-js#7025]
## Motivation
When the map is pitched, its difficult to associate the symbol name/icon with the actual point to which it is referring to.  This style would make reading maps more feasible and actually useful when it is pitched.
I have also attached  a screenshot from Google maps illustrating this.

## Design
See attached screenshot.

### Mock-Up
See attached screenshot.

![symbols_pointing_to_point_in_pitched_mode](https://user-images.githubusercontent.com/8890269/43220709-8e860580-8fff-11e8-9afb-b6be65fda217.PNG)

**Comment by andrewharvey** [comment]
I'm strongly for supporting this as part of a generalisation of `fill-extrusion` to support points and lines (these leader lines are just extruded points) together with #3993 to elevate a symbol

**Comment by ansis** [comment]
With [multi-icon symbols](https://github.com/mapbox/mapbox-gl-js/issues/4366) the leader line could be implemented as an icon. Do you think it makes sense to add this on to that issue? https://github.com/mapbox/mapbox-gl-js/issues/4366

cc @ChrisLoer who is working on multi-text right now

**Comment by peterqliu** [comment]
[check it](https://api.mapbox.com/styles/v1/peterqliu/cjk2whvfd2flk2rpp92uupgfi.html?fresh=true&title=true&access_token=pk.eyJ1IjoicGV0ZXJxbGl1IiwiYSI6ImpvZmV0UEEifQ._D4bRmVcGfJvo1wjuOpA1g#13.63/42.35596/-71.05614/-39.6/60)

<img width="921" alt="screen shot 2018-07-26 at 1 48 07 pm" src="https://user-images.githubusercontent.com/4551508/43281981-9fd03050-90da-11e8-84aa-29711a708366.png">

This was implemented in two symbol layers, that could be combined if multi-text is implemented:

- each label a text, with negative `text-offset` applied in the y axis, to effect a vertical shift upward
- each leader line also a text layer, either
  - a `text-field` of `--------` with negative `text-spacing` to remove the spaces between the hyphens (as done above)
 - a long character like `|`, though it won't let you control for length
 - producing a custom font with a line as a glyph

Note that this produces labels that are always coplanar with the display, rather than staying perpendicular with the ground plane.

>I'm strongly for supporting this as part of a generalisation of fill-extrusion to support points and lines (these leader lines are just extruded points) together with #3993 to elevate a symbol

this is tricky, because extrusions are in units of meters while text offsets are in units of px or ems

**Comment by mb12** [comment]
@ansis and @peterqliu The length of the line is a function of the pitch in G's version. It disappears when pitch==0 and placement devolves into vanilla placement without line. Would the multi icon approach model this as well? 
 You can use the following gestures in G's satellite mode to see this in action.

1. ctrl + mouse drag up/down (to change pitch) 
2. ctrl + mouse drag right/left to rotate

**Comment by ansis** [comment]
Another way to look at this would be the main icon being placed at a certain altitude `z`. And a second icon with height `z` that is rotated vertically. So, some complex combination of altitude + rotation + multi-icon might be able to do this.

**Comment by peterqliu** [comment]
>This is getting deep into hacky territory, but just as a thought exercise: we could expose pitch as a parameter to what we currently call "zoom dependent expressions" in order to scale the leader line smoothly with pitch.

I don't think this would be hacky at all! [Here it is by updating layout on the pitch event](https://bl.ocks.org/peterqliu/7147b0a605ee1d0f3412a35cb10c6c46), and an expression embed would make it even tidier. Key snippet:

```
map.on('pitch', function(){
	var ratio = map.getPitch()/60
	map
		.setLayoutProperty('leaderline', 'text-size', ratio*16)
		.setLayoutProperty('label', 'text-offset', [0, -6*ratio])
})
```

>"perpendicular to the ground plane" would basically mean the leader line length varied between 0 and 100%, as a function of pitch going from 0 to 60 degrees, right?

Yup, that's what the variable text-size is trying to simulate above ☝️, and what @mb12 seems to be getting at. Though interestingly, the text label still aligns to the screen, even when the line behaves like a real-world lamppost in the map world.

**Comment by mb12** [comment]
@peterqliu This is pretty much it. The performance is also very good.

**Comment by mb12** [comment]
@peterqliu One problem with this solution is that text-allow-overlap is set to true for both leaderline and label.  So if there are too many symbols, the map would look really cluttered. However if we turn it off, there is no way to prevent leading lines from disappearing while keeping the symbol intact and vice-versa.

**Comment by peterqliu** [comment]
>However if we turn it off, there is no way to prevent leading lines from disappearing while keeping the symbol intact and vice-versa.

@mb12 sounds like a job for the future multi-text symbol layers https://github.com/mapbox/mapbox-gl-js/issues/7025#issuecomment-408167992

**Comment by rbrundritt** [comment]
Perhaps a pitch expression would be a good option.
